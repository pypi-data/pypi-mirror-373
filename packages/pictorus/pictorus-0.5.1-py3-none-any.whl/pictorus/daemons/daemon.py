from abc import ABC, abstractmethod
from typing import Union


class Daemon(ABC):
    def create_service(self, service_name: str, description: str, bin_path: str):
        """
        Runs the daemon:
        1. Stopping the service
        2. Creating the service file
        3. Reloading the daemon
        4. Starting the service
        5. Enabling the service

        Some of these steps are optional depending on the daemon implementation
        """
        self.stop_service(service_name)
        self.create_service_file(service_name, description, bin_path)
        self.reload_daemon(service_name)
        self.start_service(service_name)
        self.enable_service(service_name)

    @abstractmethod
    def create_service_file(self, service_name: str, description: str, bin_path: str):
        """
        Create a daemon service file
        """
        pass

    @abstractmethod
    def stop_service(self, service_name: str):
        """
        Stops the daemon service
        """
        pass

    @abstractmethod
    def start_service(self, service_name: str):
        """
        Starts the daemon service
        """
        pass

    @abstractmethod
    def enable_service(self, service_name: str):
        pass

    @abstractmethod
    def reload_daemon(self, service_name: str):
        pass

    @abstractmethod
    def run(self, cmd: list, check=True, cwd=None):
        """
        The subprocess to run the daemon
        """
        pass

    @abstractmethod
    def logs(self, service_name: str, number_of_lines: Union[int, None]) -> str:
        """
        Fetches the logs of the daemon service
        """
        pass
