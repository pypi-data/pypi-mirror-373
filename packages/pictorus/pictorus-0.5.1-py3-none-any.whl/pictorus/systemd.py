import subprocess
import os

SYSTEMD_DIR = "/etc/systemd/system/"
SYSTEMCTL_BIN = "systemctl"

SERVICE_TEMPLATE = """
[Unit]
Description={description}

[Service]
ExecStart={bin_path}
Restart=always
RestartSec=5s

[Install]
WantedBy=multi-user.target
"""


def _create_service_file(service_name: str, description: str, bin_path: str):
    """Create a systemd service file"""
    service_content = SERVICE_TEMPLATE.format(description=description, bin_path=bin_path)
    with open(
        os.path.join(SYSTEMD_DIR, f"{service_name}.service"), "w", encoding="utf-8"
    ) as service_file:
        service_file.write(service_content)


def _run(cmd: list, check=True):
    subprocess.run(cmd, check=check)


def stop_service(service_name: str):
    """Stop a given service"""
    _run([SYSTEMCTL_BIN, "stop", service_name], check=False)


def start_service(service_name: str):
    """Start a given service"""
    _run([SYSTEMCTL_BIN, "start", service_name])


def enable_service(service_name: str):
    """Start a given service"""
    _run([SYSTEMCTL_BIN, "enable", service_name])


def reload_daemon():
    """Reload the systemd daemon"""
    _run([SYSTEMCTL_BIN, "daemon-reload"])


def create_service(service_name: str, description: str, bin_path: str):
    """Create and start a systemd service"""
    stop_service(service_name)
    _create_service_file(service_name, description, bin_path)
    reload_daemon()
    start_service(service_name)
    enable_service(service_name)
