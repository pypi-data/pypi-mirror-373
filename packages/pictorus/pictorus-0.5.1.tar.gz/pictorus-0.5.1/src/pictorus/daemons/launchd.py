import subprocess
import os
from typing import Union

from pictorus.daemons.daemon import Daemon

LAUNCHD_DIR = os.path.expanduser("~/Library/LaunchAgents/")
LAUNCHCTL_BIN = "launchctl"
PLIST_NAME = "us.pictor.dm.plist"
PLIST_PATH = os.path.join(LAUNCHD_DIR, PLIST_NAME)
SERVICE_LABEL = "pictorus_device_manager"

SERVICE_TEMPLATE = """<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
    <dict>
        <key>Label</key>
        <string>{service_label}</string>
        <key>Program</key>
        <string>{bin_path}</string>

        <key>StandardErrorPath</key>
        <string>/tmp/pictorus.err</string>
        <key>StandardOutPath</key>
        <string>/tmp/pictorus.out</string>
    </dict>
</plist>
"""  # noqa: E501 - Trying to format a system file


class Launchd(Daemon):
    def fill_service_template(self, bin_path):
        return SERVICE_TEMPLATE.format(service_label=SERVICE_LABEL, bin_path=bin_path)

    def create_service_file(self, service_name: str, description: str, bin_path: str):
        """Create a launchd service file"""
        service_content = self.fill_service_template(bin_path)
        with open(
            os.path.expanduser(os.path.join(LAUNCHD_DIR, PLIST_PATH)), "w", encoding="utf-8"
        ) as service_file:
            service_file.write(service_content)

    def stop_service(self, service_name: str):
        """Stop a given service"""
        self.run(
            [LAUNCHCTL_BIN, "bootout", f"gui/{os.getuid()}", PLIST_PATH],
            check=False,
            cwd=LAUNCHD_DIR,
        )

    def start_service(self, service_name: str):
        """Start a given service"""
        self.run(
            [LAUNCHCTL_BIN, "bootstrap", f"gui/{os.getuid()}", PLIST_PATH],
            check=False,
            cwd=LAUNCHD_DIR,
        )

    def enable_service(self, service_name: str):
        """Start a given service"""
        # This was obnoxious to figure out. The kickstart command references the "Label"
        # of the plist file and not the name of the file, unlike bootstrap and bootout.
        self.run(
            [LAUNCHCTL_BIN, "kickstart", "-k", f"gui/{os.getuid()}/{SERVICE_LABEL}"],
            check=False,
            cwd=LAUNCHD_DIR,
        )

    def reload_daemon(self, service_name: str):
        """Reload the launchd daemon"""

    def run(self, cmd: list, check=True, cwd=None):
        subprocess.run(cmd, check=check, cwd=cwd)

    def logs(self, service_name: str, number_of_lines: Union[int, None]) -> str:
        """Get the logs for a given service"""
        # Launchd does not have a direct command to fetch logs like systemd.
        # Instead, we can read the log files directly.
        err_log_path = "/tmp/pictorus.err"

        if os.path.exists(err_log_path):
            with open(err_log_path, "r") as err_file:
                err_logs = err_file.read().splitlines()
                if number_of_lines is not None:
                    err_logs = err_logs[-number_of_lines:]  # Get the last N lines
                err_logs = "\n".join(err_logs)  # Convert from list to string
        else:
            err_logs = "No error logs found."

        return f"{err_logs}"
