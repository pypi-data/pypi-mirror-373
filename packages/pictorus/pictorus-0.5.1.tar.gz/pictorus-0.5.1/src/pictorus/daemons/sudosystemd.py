from subprocess import run, CalledProcessError
import os
from typing import Union
from pictorus.constants import JOURNALCTL_BIN, SYSTEMCTL_BIN
from pictorus.daemons.daemon import Daemon

SYSTEMD_DIR = os.path.expanduser("/etc/systemd/system/")

SERVICE_TEMPLATE = """
[Unit]
Description={description}

[Service]
ExecStart={bin_path}
Restart=always
RestartSec=5s

[Install]
WantedBy=multi-user.target
"""


class SudoSystemd(Daemon):
    def fill_service_template(self, description, bin_path):
        return SERVICE_TEMPLATE.format(description=description, bin_path=bin_path)

    def create_service_file(self, service_name: str, description: str, bin_path: str):
        """Create a systemd service file"""

        # Create the directory if it doesn't exist
        if not os.path.exists(SYSTEMD_DIR):
            os.makedirs(SYSTEMD_DIR)

        service_content = self.fill_service_template(description, bin_path)
        with open(
            os.path.join(SYSTEMD_DIR, f"{service_name}.service"), "w", encoding="utf-8"
        ) as service_file:
            service_file.write(service_content)

    def run(self, cmd: list, check=True, cwd=None):
        run(cmd, check=check, cwd=cwd)

    def stop_service(self, service_name: str):
        """Stop a given service"""
        self.run([SYSTEMCTL_BIN, "stop", service_name], check=False)

    def start_service(self, service_name: str):
        """Start a given service"""
        self.run([SYSTEMCTL_BIN, "start", service_name])

    def enable_service(self, service_name: str):
        """Start a given service"""
        self.run([SYSTEMCTL_BIN, "enable", service_name])

    def reload_daemon(self, service_name: str = ""):
        """Reload the systemd daemon"""
        self.run([SYSTEMCTL_BIN, "daemon-reload"])

    def logs(self, service_name: str, number_of_lines: Union[int, None]) -> str:
        """Get the logs for a given service"""
        try:
            command = [JOURNALCTL_BIN, "-u", service_name, "--no-pager"]
            if number_of_lines is not None:
                command.extend(["-n", str(number_of_lines)])
            result = run(command, check=True, capture_output=True, text=True)
            return result.stdout
        except CalledProcessError as e:
            return f"Error getting logs: {e.stderr}"
