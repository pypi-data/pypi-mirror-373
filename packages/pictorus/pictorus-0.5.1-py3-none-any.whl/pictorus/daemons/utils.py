import os
import shutil

import psutil

from pictorus.daemons.daemon import Daemon
from pictorus.daemons.launchd import Launchd
from pictorus.daemons.sudosystemd import SudoSystemd
from pictorus.daemons.systemd import Systemd
from pictorus.exceptions import DaemonError
from pictorus.logging_utils import TextFormat, printf

PICTORUS_PROCESS_NAME = "pictorus-device-manager"


def is_other_device_manager_running() -> bool:
    # Pictorus device manager is running as a Python script that is in the 'cmdline' part of the
    # process. psutil.info['cmdline'] is an array that needs to be searched for the script name.
    other_pictorus_device_managers_running = False
    pid = os.getpid()
    ancestor_pids = {p.pid for p in psutil.Process(pid).parents()}
    ancestor_pids.add(pid)

    other_pids = []
    for process in psutil.process_iter(["pid", "cmdline"]):
        info = process.as_dict(attrs=["pid", "cmdline"])
        if info["cmdline"]:
            if not process.is_running():
                continue

            if process.status() in (psutil.STATUS_ZOMBIE, psutil.STATUS_DEAD):
                continue

            if info["pid"] in ancestor_pids:
                # Don't count self or ancestors as other instances
                continue

            if any(PICTORUS_PROCESS_NAME in arg for arg in info["cmdline"]):
                other_pids.append(info["pid"])
                other_pictorus_device_managers_running = True

    # One instance is ok, more than one means another instance is probably running.
    if other_pictorus_device_managers_running:
        printf(
            (
                "Pictorus device manager is already running, either as a service or manually"
                " in another terminal session. "
            ),
            TextFormat.FAIL,
        )
        printf(
            ("Process ID(s) of running pictorus-device-manager instance(s): {}".format(other_pids)),
            TextFormat.FAIL,
        )
        printf(
            (
                'Use "pictorus-cli service --stop" to stop the service if'
                ' you would like to run "pictorus-device-manager" manually from the terminal.'
                ' If you are having issues connecting to the device try "pictorus-cli service'
                ' --restart" to reset the service.'
            ),
            TextFormat.OKCYAN,
        )

    return other_pictorus_device_managers_running


def get_daemon() -> Daemon:
    """Stop the Pictorus device manager service"""
    operating_system = os.uname().sysname
    if operating_system == "Darwin":
        daemon = Launchd()

        if not shutil.which("launchctl"):
            raise DaemonError("Unable to set up device manager: launchctl not found")

    elif operating_system == "Linux":
        if os.getuid() == 0:  # Running as root
            daemon = SudoSystemd()
        else:
            daemon = Systemd()

        if not shutil.which("systemctl"):
            raise DaemonError("Unable to set up device manager: systemctl not found")

    else:
        raise DaemonError("Unsupported operating system")

    return daemon
