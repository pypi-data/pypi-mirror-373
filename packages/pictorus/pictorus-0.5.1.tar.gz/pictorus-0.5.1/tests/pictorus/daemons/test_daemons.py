from pictorus.daemons.launchd import Launchd
from pictorus.daemons.sudosystemd import SudoSystemd
from pictorus.daemons.systemd import Systemd


def test_launchd_create_service_file(snapshot):
    launchd = Launchd()
    service = launchd.fill_service_template("some_bin_path")
    assert snapshot == service


def test_systemd_create_service_file(snapshot):
    systemd = Systemd()
    service = systemd.fill_service_template("some_description", "some_bin_path")
    assert snapshot == service


def test_sudosystemd_create_service_file(snapshot):
    systemd = SudoSystemd()
    service = systemd.fill_service_template("some_description", "some_bin_path")
    assert snapshot == service
