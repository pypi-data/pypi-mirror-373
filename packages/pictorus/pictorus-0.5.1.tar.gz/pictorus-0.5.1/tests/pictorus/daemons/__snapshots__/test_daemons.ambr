# serializer version: 1
# name: test_launchd_create_service_file
  '''
  <?xml version="1.0" encoding="UTF-8"?>
  <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
  <plist version="1.0">
      <dict>
          <key>Label</key>
          <string>pictorus_device_manager</string>
          <key>Program</key>
          <string>some_bin_path</string>
  
          <key>StandardErrorPath</key>
          <string>/tmp/pictorus.err</string>
          <key>StandardOutPath</key>
          <string>/tmp/pictorus.out</string>
      </dict>
  </plist>
  
  '''
# ---
# name: test_sudosystemd_create_service_file
  '''
  
  [Unit]
  Description=some_description
  
  [Service]
  ExecStart=some_bin_path
  Restart=always
  RestartSec=5s
  
  [Install]
  WantedBy=multi-user.target
  
  '''
# ---
# name: test_systemd_create_service_file
  '''
  
  [Unit]
  Description=some_description
  
  [Service]
  ExecStart=some_bin_path
  Restart=always
  RestartSec=5s
  
  [Install]
  WantedBy=default.target
  
  '''
# ---
